File: leetcode/169. Majority Element/solution.c
// Boyer-Moore Voting Algorithm
// Time: O(n)
// Space: O(1)
int majorityElement(int* nums, int numsSize) {
    int candidate = nums[0];
    int count = 1;

    for (int i=1; i < numsSize; i++) {
        if (count == 0) {
            candidate = nums[i];
            count = 1;
        } else if (candidate == nums[i]) {
            count++;
        } else {
            count--;
        }
    }
    return candidate;
}

File: leetcode/169. Majority Element/solution.js
/**
 * @param {number[]} nums
 * @return {number}
 */
// Boyer-Moore Voting Algorithm
// Time: O(n)
// Space: O(1)
var majorityElement = function(nums) {
    let candidate = nums[0];
    let count = 1;

    for (let i = 1; i < nums.length; i++) {
        if (count === 0) {
            candidate = nums[i];
            count = 1;
        } else if (nums[i] === candidate) {
            count++;
        } else {
            count--;
        }
    }
    return candidate
};
